### PR TITLE
Allow actions based on the account type in Polkadot

### LINKED ISSUE #4017: Allow actions based on the account type in Polkadot
In Polkadot there are 2 account types: Stash and Controller. In Lunie we set both to be the same. But certain actions need certain account types. If not signing will fail.

TO DO:
1. PR
- [x] identify account type when logging into a Polkadot network
2. PR
- [ ] if Stash account and not the controller account for that account, disable delegation to validators (keep bonding enabled, disable validator selection)
- [ ] if Stash account and not the controller account for that account, show a warning "For staking to validators you need to sign in with your controller account ACCOUNT HERE. You can still stake tokens."
- [ ] if Controller account show a warning on the portfolio: "this is a controller account, if you want to see your validators, sign in with your stash account."
3. PR
- [ ] check batch message construction to not include the wrong tx

### PR DESCRIPTION
**Pull request by mariopino**
Closes #4017 

**Description:**

Allow actions based on the account type in Polkadot:

- [x] Get address role from API at sign-in
- [x] Display role in AppMenu
- [x] If Stash account (!= controller): Enable bonding, disable validator selection.
- [x] If Stash account (!= controller) show a warning: This is a stash account, you can increase the amount to stake but you need to sign in with your controller account to set or change your validators.
- [x] If Controller account  (!= stash): Disable bonding, enable validator selection.
- [x] If Controller account (!= stash) show a warning: This is a controller account, you can set or change your validators but to increase the amount to stake you need to sign in with your stash account.."
- [x] Modify batch message construction to not include the wrong tx
- [ ] For controllers we can display the delegations with this API PR: https://github.com/luniehq/lunie-api/pull/775

image

image

Thank you! 🚀

---

For contributor:

- [x] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

File: src/components/common/AppMenu.vue [hunk 1]
```diff
@@ -3,7 +3,14 @@
     <div v-if="session.signedIn" class="user-box">
       <div class="user-box-address">
         <div>
-          <h3>Your Address</h3>
+          <h3
+            v-if="
+              session.addressRole && session.addressRole !== `stash/controller`
```
Comment (Bitcoinera) [inline src/components/common/AppMenu.vue:8]: Is this accounting for non-Polkadot addresses? I mean, do non-Polkadot addresses have `stash-controller` as their accountRole?

**Comment by Bitcoinera** [comment]
I just saw this one 😆 on which branch should we work now? This is getting quite confusing... 🤷

File: src/vuex/modules/session.js [hunk 2]
```diff
@@ -1,272 +1,264 @@
...
-      state,
-      mutations,
-      actions
-    }
+  return {
+    state,
+    mutations,
+    actions
   }
+}
```
Comment (Bitcoinera) [inline src/vuex/modules/session.js:261]: We really do need to fix the linting thing this file here is unreadable. Cannot really see the changes made 🤷

File: src/components/common/AppMenu.vue (hunk 1 above)
Comment (Bitcoinera) [reply src/components/common/AppMenu.vue:8]: I think, by looking at session.js, that it would be `undefined` on those cases

**Comment by faboweb** [comment]
Please do:
- for a stash account disable the delegation/undelegation from field in the form and show the inline notification as mentioned
- for a controller account disable the amount field in delegation/undelegation and show the inline notification mentioned

File: src/components/staking/DelegationsOverview.vue [hunk 3]
```diff
@@ -1,5 +1,29 @@
...
+      </TmDataMsg>
+      <TmDataMsg
+        v-else-if="session.addressRole === `stash`"
+        icon="sentiment_dissatisfied"
+      >
+        <div slot="title">
+          This is a stash account
+        </div>
+        <div slot="subtitle">
+          Staking actions are disabled. You can still send tokens.
```
Comment (faboweb) [inline src/components/staking/DelegationsOverview.vue:23]: this is the same as the other

File: package.json [hunk 4]
```diff
@@ -140,7 +140,8 @@
     "minimist": ">=1.2.3",
     "yargs-parser": ">=18.1.2",
     "@vue/cli-service/**/yargs-parser": "^13.1.2",
-    "@hapi/hoek": ">=9.0.3"
+    "@hapi/hoek": ">=9.0.3",
+    "kind-of": ">=6.0.3"
```
Comment (faboweb) [inline package.json:144]: what is that?

Comment (mariopino) [reply package.json:144]: Fix for a security vuln.

**Comment by mariopino** [comment]
> Please do:
> 
> * for a stash account disable the delegation/undelegation from field in the form and show the inline notification as mentioned
> * for a controller account disable the amount field in delegation/undelegation and show the inline notification mentioned

Working on it ...

File: src/vuex/modules/session.js [hunk 5]
```diff
@@ -1,272 +1,264 @@
...
+      if (currentNetwork.network_type === "polkadot") {
+        await dispatch(`checkAddressRole`, {
           address,
-          sessionType,
-          networkId: currentNetwork.id
-        })
-        const addresses = state.addresses
-        dispatch(`persistAddresses`, {
-          addresses
+          currentNetwork
```
Comment (Bitcoinera) [inline src/vuex/modules/session.js:160]:
```suggestion
          networkId
```

File: src/vuex/modules/session.js [hunk 6]
```diff
@@ -1,272 +1,264 @@
...
+    },
+    setPreferredCurrency({ state, dispatch }, currency) {
+      state.preferredCurrency = currency
+      dispatch(`storeLocalPreferences`)
+    },
+    /* istanbul ignore next */
+    async checkAddressRole({ commit }, { address, currentNetwork }) {
+      const { data } = await apollo.query({
+        query: AddressRole,
+        variables: { networkId: currentNetwork.id, address: address },
```
Comment (Bitcoinera) [inline src/vuex/modules/session.js:249]:
```suggestion
        variables: { networkId, address },
```

File: src/vuex/modules/session.js [hunk 7]
```diff
@@ -1,272 +1,264 @@
...
+        state.externals.anonymize()
+        console.log(`Analytics collection has been disabled`)
       }
+    },
+    setPreferredCurrency({ state, dispatch }, currency) {
+      state.preferredCurrency = currency
+      dispatch(`storeLocalPreferences`)
+    },
+    /* istanbul ignore next */
+    async checkAddressRole({ commit }, { address, currentNetwork }) {
```
Comment (Bitcoinera) [inline src/vuex/modules/session.js:246]:
```suggestion
    async checkAddressRole({ commit }, { address, networkId}) {
```

File: src/components/staking/DelegationsOverview.vue [hunk 8]
```diff
@@ -1,5 +1,29 @@
 <template>
   <div class="table-container">
+    <template v-if="session.addressRole">
```
Comment (faboweb) [inline src/components/staking/DelegationsOverview.vue:3]: remove?

Comment (mariopino) [reply src/components/staking/DelegationsOverview.vue:3]: I think yep

**Comment by faboweb** [comment]
1 thing: Can you set the `To` input to empty. So the validator address doesn't show?

File: src/vuex/modules/session.js [hunk 9]
```diff
@@ -249,6 +249,7 @@ export default ({ apollo }) => {
         variables: { networkId, address },
         fetchPolicy: "network-only"
       })
+      console.log(data)
```
Comment (faboweb) [inline src/vuex/modules/session.js:252]:
```suggestion
```

**Comment by faboweb** [comment]
Amount 0 is not working on staking

**Comment by faboweb** [comment]
> Amount 0 is not working on staking

fixed

**Comment by faboweb** [comment]
Merged so it can be used in #4016. let's use follow up Prs if anything is not working

File: src/components/staking/DelegationsOverview.vue [hunk 10]
```diff
@@ -52,7 +48,8 @@ export default {
     delegationsLoaded: false,
   }),
   computed: {
-    ...mapGetters(["address", `network`, `networks`]),
+    ...mapState([`session`]),
```
Comment (faboweb) [inline src/components/staking/DelegationsOverview.vue:51]: this is now actually not needed

**Pull request merged**